### PR TITLE
Task listing add definition of done test

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java [hunk 1]
```diff
@@ -176,4 +178,43 @@ void givenTwoTaskManagerATaskSubmittedOnOneCouldBeRunOnTheOther() throws Interru
...
+        TaskId taskId1 = taskManager1.submit(new CompletedTask());
+        TaskId taskId2 = taskManager2.submit(new CompletedTask());
+
+        ImmutableSet<TaskId> expectedTaskIds = ImmutableSet.of(taskId1, taskId2);
+        int expectedNumberOfTasks = 2;
+
+        Awaitility.await()
+            .atMost(Duration.ONE_SECOND)
+            .pollInterval(100L, TimeUnit.MILLISECONDS)
+            .until(() -> {
```
Comment (chibenwa) [inline server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: I think you can achieve a more readable output by relying on "untilAsserted" + dump regular assertj statements...

**Comment by rouazana** [comment]
```
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] to be an instance of:
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425]   <org.apache.james.StartUpChecksPerformer.StartUpChecksException>
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] but was:
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425]   <"com.google.inject.CreationException: Unable to create injector, see the following errors:
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] 
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] 1) A binding to org.apache.james.task.eventsourcing.Hostname was already configured at org.apache.james.modules.server.HostnameModule.configure(HostnameModule.java:41) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.CommonServicesModule -> org.apache.james.modules.server.TaskManagerModule -> org.apache.james.modules.server.HostnameModule).
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425]   at org.apache.james.modules.server.HostnameModule.configure(HostnameModule.java:40) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.CommonServicesModule -> org.apache.james.modules.server.TaskManagerModule -> org.apache.james.modules.server.HostnameModule)
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] 
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] 2) Could not find a suitable constructor in org.apache.james.task.MemoryTaskManager. Classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425]   at org.apache.james.task.MemoryTaskManager.class(MemoryTaskManager.java:85)
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425]   at org.apache.james.modules.server.TaskManagerModule.configure(TaskManagerModule.java:32) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.CommonServicesModule -> org.apache.james.modules.server.TaskManagerModule)
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] 
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] 3) Could not find a suitable constructor in org.apache.james.task.eventsourcing.Hostname. Classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425]   at org.apache.james.task.eventsourcing.Hostname.class(Events.scala:29)
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425]   at org.apache.james.modules.server.HostnameModule.configure(HostnameModule.java:40) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.CommonServicesModule -> org.apache.james.modules.server.TaskManagerModule -> org.apache.james.modules.server.HostnameModule)
[dd775f6876c7bdeebe3e9222adb6f94c9d5ca425] 
```

File: server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java [hunk 2]
```diff
@@ -176,4 +177,44 @@ void givenTwoTaskManagerATaskSubmittedOnOneCouldBeRunOnTheOther() throws Interru
...
+
+        TaskId taskId1 = taskManager1.submit(new CompletedTask());
+        TaskId taskId2 = taskManager2.submit(new CompletedTask());
+
+        int expectedNumberOfTasks = 2;
+
+        Awaitility.await()
+            .atMost(Duration.ONE_SECOND)
+            .pollInterval(100L, TimeUnit.MILLISECONDS)
+            .untilAsserted(() -> {
```
Comment (mbaechler) [inline server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: it's usually a bad idea to put the actual assertion into an until. Usually, we wait for some events then we do the assertions.

Comment (remk) [reply server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: At first i used a condition into an until. But without assertion after that.
From a previous comment i refactored it with untilAsserted.
I don't see the point in rewriting it in this particular case. As the conditions needed  to do the assertions correspond exactly to the assertions.

Comment (mbaechler) [reply server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: Can you tell me what's the output of the test when it fails?

Comment (remk) [reply server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: Yes for example if i make the test fail on the assertion comparing the size of the list.
I have the line and actual/expected value.

```
Expecting:
 <2>
to be equal to:
 <3>
but was not.
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at org.apache.james.task.eventsourcing.distributed.DistributedTaskManagerTest.lambda$givenTwoTaskManagerAndATaskRanPerTaskManagerListingThemOnEachShouldShowBothTasks$5(DistributedTaskManagerTest.java:215)
```

File: server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java [hunk 3]
```diff
@@ -178,4 +179,44 @@ void givenTwoTaskManagerATaskSubmittedOnOneCouldBeRunOnTheOther() throws Interru
...
+            TaskId taskId1 = taskManager1.submit(new CompletedTask());
+            TaskId taskId2 = taskManager2.submit(new CompletedTask());
+
+            int expectedNumberOfTasks = 2;
+
+            Awaitility.await()
+                .atMost(Duration.ONE_SECOND)
+                .pollInterval(100L, TimeUnit.MILLISECONDS)
+                .untilAsserted(() -> {
+                    List<TaskExecutionDetails> listOnFirstTaskManager = taskManager1.list();
```
Comment (chibenwa) [inline server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: listOnTaskManager1

File: server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java [hunk 4]
```diff
@@ -178,4 +179,44 @@ void givenTwoTaskManagerATaskSubmittedOnOneCouldBeRunOnTheOther() throws Interru
...
+                    List<TaskExecutionDetails> listOnFirstTaskManager2 = taskManager2.list();
```
Comment (chibenwa) [inline server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: s/listOnFirstTaskManager2/listOnTaskManager2

it 's not the first, right?

File: server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java [hunk 5]
```diff
@@ -178,4 +179,44 @@ void givenTwoTaskManagerATaskSubmittedOnOneCouldBeRunOnTheOther() throws Interru
...
+                    List<TaskId> actualTaskIds = extractIds(listOnFirstTaskManager);
+
+                    assertThat(listOnFirstTaskManager.size()).isEqualTo(expectedNumberOfTasks);
```
Comment (chibenwa) [inline server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: Let's use `.hasSize` and rely on more fluent asserts?

```
assertThat(listOnTaskManager1)
    .hasSize(2)
    .isEqualTo(listOnTaskManager2)
    .allSatisfies(taskExecutionDetails -> assertThat(taskExecutionDetails.getStatus()).isEqualTo(COMPLETED))
    .extracting(TaskExecutionDetails::getTaskId)
    .containsExactlyInAnyOrder(taskId1, taskId2);
```

File: server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java [hunk 6]
```diff
@@ -178,4 +179,36 @@ void givenTwoTaskManagerATaskSubmittedOnOneCouldBeRunOnTheOther() throws Interru
             }
         }
     }
+
+    private List<TaskId> extractIds(List<TaskExecutionDetails> tasks) {
```
Comment (chibenwa) [inline server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: This method is no longer needed

Comment (remk) [reply server/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: Yeah i pushed in the meantime...

**Comment by blackheaven** [comment]
fetch error

test this please

**Comment by remk** [comment]
```
[fcd462d73c185b511373f6e700751f1a20de33c4] java.lang.InterruptedException: null
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2056)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2090)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at java.base/java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:433)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at reactor.core.publisher.MonoCallable.subscribe(MonoCallable.java:56)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at reactor.core.publisher.Mono.subscribe(Mono.java:3852)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at reactor.core.publisher.FluxRepeatPredicate$RepeatPredicateSubscriber.resubscribe(FluxRepeatPredicate.java:112)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at reactor.core.publisher.MonoRepeatPredicate.subscribe(MonoRepeatPredicate.java:47)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at reactor.core.publisher.Flux.subscribe(Flux.java:7921)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.run(FluxSubscribeOn.java:194)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[fcd462d73c185b511373f6e700751f1a20de33c4] 	at java.base/java.lang.Thread.run(Thread.java:834)
[fcd462d73c185b511373f6e700751f1a20de33c4] [INFO] Tests run: 28, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.343 s - in org.apache.james.jmap.memory.MemoryJmapAuthenticationTest
[fcd462d73c185b511373f6e700751f1a20de33c4] [INFO] 
[fcd462d73c185b511373f6e700751f1a20de33c4] [INFO] Results:
[fcd462d73c185b511373f6e700751f1a20de33c4] [INFO] 
[fcd462d73c185b511373f6e700751f1a20de33c4] [ERROR] Errors: 
[fcd462d73c185b511373f6e700751f1a20de33c4] [ERROR]   MemorySpamAssassinContractTest.imapCopiesToSpamMailboxShouldBeConsideredAsSpam » ConditionTimeout
```

test this please

**Comment by blackheaven** [comment]
test this please

**Comment by blackheaven** [comment]
```
[8591d35bacf1d7b867d722bb021b09c34d439b6f] Expected: '\* SEARCH 1 2 3 4'
[8591d35bacf1d7b867d722bb021b09c34d439b6f] Actual   : '* SEARCH'
```

test this please

**Comment by remk** [comment]
` CassandraMailRepositoryTest.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult `

**Comment by remk** [comment]
test  this please

**Comment by blackheaven** [comment]
```
[ERROR]   CassandraMailRepositoryTest.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult » Execution
```

test this please

**Comment by mbaechler** [comment]
merged

**Pull request closed without merging**